### ISSUE TITLE
Add `[disallow-non-exhaustive-match]` check

### ISSUE THREAD

**Issue by johnthagen** [python/mypy#13597]
**Feature**

Add a new off-by-default check, `[disallow-non-exhaustive-match]`, that enforces at type check time that no `match` statements implicitly fall through their bottom.

**Pitch**

The (superseded) [PEP 622](https://peps.python.org/pep-0622/#exhaustiveness-checks) has a great discussion on the motivation for this feature with several examples.

For brevity, below we list a very simple example demonstrating the kinds of correctness checks that motivate this feature request.

Consider we have the following code:

```py
from enum import Enum, auto

class Color(Enum):
    Red = auto()
    Green = auto()

def print_color(color: Color) -> None:
    match color:
        case Color.Red:
            print("red")
        case Color.Green:
            print("green")
```

Now, consider in the future another `Color` variant is added:

```py
from enum import Enum, auto

class Color(Enum):
    Red = auto()
    Green = auto()
    Blue = auto()

def print_color(color: Color) -> None:
    match color:
        case Color.Red:
            print("red")
        case Color.Green:
            print("green")
```

Oops! We forgot to add another case to `print_color`'s `match`, `Color.Blue` is not handled, and this bug slips though into runtime.

Even if `mypy` is configured in `strict` mode, this bug isn't caught (because Python's default behavior is to allow `match` fall through).

```toml
[tool.mypy]
strict = true
```

```shell
$ mypy main.py
Success: no issues found in 1 source file
```

So how does a programmer defend against this bug? Currently, the best solution is to sprinkle [`assert_never`](https://docs.python.org/3.11/library/typing.html#typing.assert_never) clauses into every `match`.

```py
from enum import Enum, auto

from typing_extensions import assert_never


class Color(Enum):
    Red = auto()
    Green = auto()
    Blue = auto()


def print_color(color: Color) -> None:
    match color:
        case Color.Red:
            print("red")
        case Color.Green:
            print("green")
        case _ as unreachable:
            assert_never(unreachable)
```

This will catch the error:

```shell
$ mypy main.py
main.py:19: error: Argument 1 to "assert_never" has incompatible type "Literal[Color.Blue]"; expected "NoReturn"
```

This boilerplate requires the programmer to:

1. If running on Python <3.11, add `typing_extensions` to their dependencies
2. For each module they are using `match`, import `assert_never`
3. For each `match` statement, add the `case _` clause

This hurts the ergonomics of trying to pervasively enforce this kind of correctness.

This also requires the programmer not to forget to add this boilerplate to every `match` clause, which can be error prone, especially for large projects wanting to guarantee correctness.

If using [`coverage`](https://coverage.readthedocs.io/en/6.4.4/excluding.html), the boilerplate increases further as there will be no reasonable way to add tests to cover the extra unreachable `case`s. An extra `# pragma: no cover` comment must be added to every unreachable `case`:

```py
def print_color(color: Color) -> None:
    match color:
        case Color.Red:
            print("red")
        case Color.Green:
            print("green")
        case _ as unreachable:  # pragma: no cover
            assert_never(unreachable)
```

**Other Languages**

Other languages with pattern matching, such as [Rust](https://doc.rust-lang.org/rust-by-example/flow_control/match.html), enforce exhaustive pattern matching, and it has been observed to have a positive effect on program correctness.

**Comment by ethanhs** [comment]
I actually would prefer this to be on by default. The superseding PEP 634 says:

> For cases such as sealed classes and enums, where the patterns are all known to be members of a discrete set, [static checkers](https://peps.python.org/pep-0622/#static-checkers) can warn about missing patterns.

In addition, there are already so many flags and I think this is a case where it is reasonable to diverge from runtime.

**Comment by hauntsaninja** [comment]
Note that mypy will already complain about code like the following, so this shouldn't be a hard check to add:
```
from enum import Enum, auto

class Color(Enum):
    Red = auto()
    Green = auto()
    Blue = auto()

def print_color(color: Color) -> int:
    match color:
        case Color.Red:
            return 4
        case Color.Green:
            return 5
```
Take a look at https://github.com/python/mypy/pull/12267 if you're interested in implementing this.

For what it's worth:
- pyright rejected enforcing exhaustiveness: https://github.com/microsoft/pyright/issues/2569#issuecomment-969936687
- Guido was also hesitant about enforcing exhaustiveness, but mentioned he could see value in having this per-module (and mypy master now supports per-module error codes): https://github.com/python/mypy/pull/10191#issuecomment-1016608286

**Comment by hauntsaninja** [comment]
(A little off topic, but if you have open source code that makes interesting use of typing features, pattern matching, or has surfaced mypy usability issues for you previously, let me know, and I can add it tohttps://github.com/hauntsaninja/mypy_primer )

**Comment by johnthagen** [comment]
> I actually would prefer this to be on by default.

@ethanhs I would too. I clarified the initial post to express this. I originally posted off-by-default to try to avoid controversy if the mypy team feels we shouldn't diverge from the runtime behavior. But I agree with everything you said above.

**Comment by bskinn** [comment]
TypeScript also is equipped to evaluate its `[key in <type>]` construction exhaustively, I believe by default.

The following shows a `Property '[E_Color.Blue]' is missing in type '{ red: string; green: string; }' but required in type '{ red: string; green: string; blue: string; }'` error in one of my projects:

```
export enum E_Color {
  Red = 'red',
  Green = 'green',
  Blue = 'blue',
}

export const colorsUpper: {[key in E_Color]: string} = {  // <-- Error
  [E_Color.Red]: 'RED',
  [E_Color.Green]: 'GREEN',
}
```

**Comment by 9999years** [comment]
I was pretty shocked to discover that this behavior isn't implemented.

**Comment by stevapple** [comment]
> OCaml and Swift as well, Haskell -- or at least ghc -- stands out for not having exhaustive checking by default, and IIRC it's because the type system is sufficiently complex that it's a hard problem in that specific language, as well as the relevant GHC code being pretty old and crufty and not having a real champion / owner.

Swift doesn't seem to be on this list. In Swift you'll always need a `default:` block (identical to `case _:` in Python) if you're not covering all the cases. What's not enabled by default is the enforcement of `@unknown default:`, i.e. always requiring a fallback for non-frozen `enum`s in case a new `case` is added afterwards.

**Comment by bobot** [comment]
> OCaml and Swift as well, Haskell -- or at least ghc -- stands out for not having exhaustive checking by default,

:open_mouth: OCaml, check by default the exhaustiveness and [many other properties](https://ocaml.org/docs/common-errors#pattern-matching-warnings-and-errors) in patterns, and Haskell check exhaustiveness with `-Wall`.

**Comment by A5rocks** [comment]
This is now implemented as the `exhaustive-match` error, disabled by default. Hopefully we can enable it by default in strict modes but that's for another day.

**Comment by hauntsaninja** [comment]
For copy-pasting purposes: `--enable-error-code exhaustive-match`

https://mypy.readthedocs.io/en/latest/error_code_list2.html#check-that-match-statements-match-exhaustively-match-exhaustive

Thanks to Don-Burns in https://github.com/python/mypy/pull/19144

**Comment by johnthagen** [comment]
For copy-pasting purposes to `pyproject.toml`:

```toml
[tool.mypy]
enable_error_code = ["exhaustive-match"]
```

Also, @hauntsaninja, is there a typo in the docs, or is the error code inconsistent with the CLI flag?

https://mypy.readthedocs.io/en/latest/error_code_list2.html#check-that-match-statements-match-exhaustively-match-exhaustive

The title says `match-exhaustive` but the examples use `exhaustive-match`.

**Comment by brianschubert** [comment]
> The title says match-exhaustive but the examples use exhaustive-match.

Yeah, that looks like a typo. Feel free to send a PR!

**Comment by johnthagen** [comment]
@brianschubert 

- #19276
